Added blog post semantic search in postgres in 15 minutes (#1535)
postgresml/postgresml: pgml-cms/blog/semantic-search-in-postgres-in-15-minutes.md

---
description: >-
  Learn how to implement semantic search in PostgreSQL with nothing but SQL.
featured: true
tags: ["Engineering"]
---

# Semantic Search in Postgres in 15 Minutes

<div align="left">

<figure><img src=".gitbook/assets/silas.jpg" alt="Author" width="125"><figcaption></figcaption></figure>

</div>

Silas Marvin

June 15, 2024

## What is and is not semantic search

Semantic search uses machine learning to understand the meaning of text by converting it into numerical vectors, allowing for more accurate and context-aware search results.

When users are unsure of the exact terms to search for, semantic search can uncover relevant information that traditional keyword searches might miss. This capability is particularly valuable for discovering content based on the intent and context of the search query, rather than relying solely on precise word matches.

> It is not a complete replacement for full-text search. In many cases, full-text search can outperform semantic search. Specifically, if a user knows the exact phrase they want to match in a document, full-text search is faster and guaranteed to return the correct result, whereas semantic search is only likely to return the correct result. Full-text search and semantic search can be combined to create powerful and robust search systems.

Reviewer: I'd also go a bit further in this section to mention re-ranking is applicable to both, and as long as you need keyword search, you need classical support for entity recognition, and graph associations, which can also improve semantic matching.

> Semantic search is not just for machine learning engineers. The system behind semantic search is relatively easy to implement, and thanks to new Postgres extensions like _pgml_ and _pgvector_, it is readily available to SQL developers. Just as modern SQL developers are expected to be familiar with and capable of implementing full-text search, they will soon be expected to implement semantic search as well.

Reviewer: ```suggestion
Semantic search is not just for machine learning engineers. The system behind semantic search is relatively easy to implement, and thanks to new Postgres extensions like `pgml` and `pgvector`, it is readily available to SQL developers. Just as modern SQL developers are expected to be familiar with and capable of implementing full-text search, they will soon be expected to implement semantic search as well.
```

## Embeddings 101

Semantic search is powered by embeddings. To understand how semantic search works, we must have a basic understanding of embeddings. 

> Embeddings are vectors. Given some text and some embedding model, we can convert text to vectors:

Reviewer: more programmers are familiar with arrays than vectors if you're looking for a common explanation.

```

!!!

!!!

We used the [pgml.embed](/docs/api/sql-extension/pgml.embed) PostresML function to generate an embedding of the sentence "Generating embeddings in Postgres is fun!" using the [mixedbread-ai/mxbai-embed-large-v1](https://huggingface.co/mixedbread-ai/mxbai-embed-large-v1) model from mixedbread.ai.

> The output size of the vector varies per model, and in _mxbai-embed-large-v1_ outputs vectors with 1024 dimensions: each vector contains 1024 floating point numbers. 

Reviewer: ```suggestion
The output size of the vector varies per model, and in `mxbai-embed-large-v1` outputs vectors with 1024 dimensions: each vector contains 1024 floating point numbers. 
```

I think we should use backticks for all identifiers, rather than italics.

Reviewer: I agree that we could make some design improvements here. It'd be good to bring this up with the design team. We should still consistently use semantic markup, so that when the style is fixed, we get the benefits.

The vector this model outputs is not random. It is designed to capture the semantic meaning of the text. What this really means, is that sentences which are closer together in meaning will be closer together in vector space. 

> Let’s look at a more simple example. Let's assume we have a model called _simple-embedding-model_, and it outputs vectors with only 2 dimensions. Let’s embed the following three phrases: "I like Postgres", "I like SQL" and "Rust is the best": 

Reviewer: ```suggestion
Let’s look at a more simple example. Let's assume we have a model called `simple-embedding-model`, and it outputs vectors with only 2 dimensions. Let’s embed the following three phrases: "I like Postgres", "I like SQL" and "Rust is the best": 
```

!!! generic

!!! code_block

```postgresql
SELECT pgml.embed('simple-embedding-model', 'I like Postgres') AS embedding;

SELECT pgml.embed('simple-embedding-model', 'I like SQL') AS embedding;

SELECT pgml.embed('simple-embedding-model', 'Rust is the best') AS embedding;
```

!!!

!!! results

```text
embedding for 'I like Postgres'
---------
[0.1, 0.2]

embedding for 'I like SQL'
---------
[0.12, 0.25]

embedding for 'Rust is the best'
---------
[-0.8, -0.9]
```

!!!

!!!

> You'll notice how similar the vectors produced by the text "I like Postgres" and "I like SQL" are compared to "Rust is the best".  This is a artificial example, but the same idea holds true when translating to real models like _mixedbread-ai/mxbai-embed-large-v1_. 

Reviewer: ```suggestion
You'll notice how similar the vectors produced by the text "I like Postgres" and "I like SQL" are compared to "Rust is the best".  This is a artificial example, but the same idea holds true when translating to real models like `mixedbread-ai/mxbai-embed-large-v1`. 
```

## What does it mean to be "close"?

We can use the idea that text that is more similar in meaning will be closer together in the vector space to build our semantic search engine. 

For instance let’s say that we have the following documents:

| Document ID | Document text |
-----|----------|
| 1 | The pgml.transform function is a PostgreSQL function for calling LLMs in the database. |
| 2 | I think tomatoes are incredible on burgers. |

and a user is looking for the answer to the question: "What is the pgml.transform function?". If we embed the search query and all of the documents using a model like _mixedbread-ai/mxbai-embed-large-v1_, we can compare the query embedding to all of the document embeddings, and select the document that has the closest embedding in vector space, and therefore in meaning, to the to the answer. 

These are big embeddings, so we can’t simply estimate which one is closest. So, how do we actually measure the similarity (distance) between different vectors?

> _pgvector_ as of this writing supports four different measurements of vector similarity:

Reviewer: ```suggestion
`pgvector` as of this writing supports four different measurements of vector similarity:
```

- L2 distance
- (negative) inner product
- cosine distance
- L1 distance

> For most use cases we recommend using the cosine distance as defined by the formula:

Reviewer: You may be able to shorten this section with a link to our docs on distance functions.

Author: I absolutely can, but do we want to? I was thinking this would post would give enough of an intro that after going through
it they would feel confident enough to reproduce this on their own test project, or rather, use this as a guide and copy the steps we did without too much hand waving. If the length of this section takes away from that goal I will trim it down!

<figure><img src=".gitbook/assets/cosine_similarity.png" alt="cosine similarity formula"><figcaption></figcaption></figure>

where A and B are two vectors. 

> This is a somewhat confusing formula but luckily  _pgvector_ provides an operator that computes the cosine distance for us:

Reviewer: ```suggestion
This is a somewhat confusing formula but luckily  `pgvector` provides an operator that computes the cosine distance for us:
```

Going back to our search example, we can compute the cosine distance between our query embedding and our documents:

!!! generic

!!! code_block

```postgresql
SELECT pgml.embed(
  'mixedbread-ai/mxbai-embed-large-v1',
  'What is the pgml.transform function?'
)::vector
  <=>
pgml.embed(
  'mixedbread-ai/mxbai-embed-large-v1',
  'The pgml.transform function is a PostgreSQL function for calling LLMs in the database.'
)::vector AS cosine_distance;

SELECT pgml.embed(
  'mixedbread-ai/mxbai-embed-large-v1',
  'What is the pgml.transform function?'
)::vector
  <=>
pgml.embed(
  'mixedbread-ai/mxbai-embed-large-v1',
  'I think tomatoes are incredible on burgers.'
)::vector AS cosine_distance;
```

!!!

!!! results

```text
cosine_distance   
--------------------
 0.1114425936213167

cosine_distance   
--------------------
 0.7328613577628744
```

!!!

!!!

> You'll notice that the distance between "What is the pgml.transform function?" and "The pgml.transform function is a PostgreSQL function for calling LLMs in the database." is much smaller than the cosine distance between "What is the pgml.transform function?" and "I think tomatoes are incredible on burgers".

Reviewer: It's probably worth mentioning "out of sample" tokens, since they are rife on domain specific topics like this.

Author: I'm not sure what that is?

## Making it fast!

It is inefficient to compute embeddings for all the documents every time we search the dataset. Instead, we should embed our documents once and search against precomputed embeddings. 

> _pgvector_ provides us with the `vector` data type for storing embeddings in regular PostgreSQL tables:

Reviewer: ```suggestion
`pgvector` provides us with the `vector` data type for storing embeddings in regular PostgreSQL tables:
```

          '{"prompt": "Represent this sentence for searching relevant passages: "}'
        )::vector embedding
)
SELECT
    text,
    (
        SELECT
          embedding
        FROM query_embedding
    ) <=> text_and_embeddings.embedding cosine_distance
FROM
  text_and_embeddings
ORDER BY cosine_distance
LIMIT 1;
```

!!!

!!! results

```
                                          text                                          |   cosine_distance   
----------------------------------------------------------------------------------------+---------------------
 The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
```

!!!

!!!

This query is fast for now, but as we add more data to the table, it will slow down because we have not indexed the embedding column. 

Let's demonstrate this by inserting 100,000 additional embeddings:

!!! generic

!!! code_block

```postgresql
INSERT INTO text_and_embeddings (text, embedding) 
SELECT
  md5(random()::text),
  pgml.embed(
    'mixedbread-ai/mxbai-embed-large-v1',
    md5(random()::text)
  ) 
FROM generate_series(1, 100000);
```

!!!

!!!

> Now trying our search engine again:

Reviewer: This would be better illustrated with a graph at 10, 100, 1000, 10000, 100k, 1M. Show that Big O.

!!! generic

!!! code_block time="105.917 ms"

```postgresql
WITH embedded_query AS (
    SELECT
        pgml.embed('mixedbread-ai/mxbai-embed-large-v1', 'What is the pgml.transform function?', '{"prompt": "Represent this sentence for searching relevant passages: "}')::vector embedding
)
SELECT
    text,
    (
        SELECT
            embedding
        FROM embedded_query) <=> text_and_embeddings.embedding cosine_distance
FROM
  text_and_embeddings
ORDER BY cosine_distance
LIMIT 1;
```

!!!

!!! results

```
                                          text                                          |   cosine_distance   
----------------------------------------------------------------------------------------+---------------------
 The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
```

!!!

!!!

> This somewhat less than ideal performance can be fixed by indexing the embedding column. There are two types of indexes available in _pgvector_: IVFFlat and HNSW.

Reviewer: ```suggestion
This somewhat less than ideal performance can be fixed by indexing the embedding column. There are two types of indexes available in `pgvector`: IVFFlat and HNSW.
```

!!! generic

!!! code_block time="4989.398 ms"

```postgresql
CREATE INDEX ON text_and_embeddings
USING ivfflat (embedding vector_cosine_ops)
WITH (lists = 10);
```

!!!

!!!

and search again, we would get much better performance:

!!! generic

!!! code_block time="29.191 ms"

```postgresql
WITH embedded_query AS (
    SELECT
        pgml.embed('mixedbread-ai/mxbai-embed-large-v1', 'What is the pgml.transform function?', '{"prompt": "Represent this sentence for searching relevant passages: "}')::vector embedding
)
SELECT
    text,
    (
        SELECT
            embedding
        FROM embedded_query) <=> text_and_embeddings.embedding cosine_distance
FROM
  text_and_embeddings
ORDER BY cosine_distance
LIMIT 1;
```

!!!

!!! results

```
                                          text                                          |   cosine_distance   
----------------------------------------------------------------------------------------+---------------------
 The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
```

!!!

!!!

> We can see it is a massive speedup because we are only searching over 1/10th of the original vectors! 

Reviewer: ```suggestion
We can see it is a massive speedup because we are only comparing our input to 1/10th of the original vectors, instead of all of them! 
```

> HNSW indexes are a bit more complicated. It is essentially a graph with edges linked by proximity in vector space. For more information you can check out this [write-up](https://www.pinecone.io/learn/series/faiss/hnsw/). 

Reviewer: Link to wikipedia if we have to link out, but links out usually end a session.

> HNSW indexes typically have better and faster recall but require more compute when adding new vectors. That being said, we recommend using HNSW indexes for most use cases.

Reviewer: ```suggestion
HNSW indexes typically have better and faster recall but require more compute when adding new vectors. That being said, we recommend using HNSW indexes for most use cases where writes are less frequent than reads.
```

          'What is the pgml.transform function?',
          '{"prompt": "Represent this sentence for searching relevant passages: "}'
        )::vector embedding
)
SELECT
    text,
    (
        SELECT
          embedding
        FROM embedded_query
    ) <=> text_and_embeddings.embedding cosine_distance
FROM
  text_and_embeddings
ORDER BY cosine_distance
LIMIT 1;
```

!!!

!!! results

```
                                          text                                          |   cosine_distance   
----------------------------------------------------------------------------------------+---------------------
 The pgml.transform function is a PostgreSQL function for calling LLMs in the database. | 0.13467974993681486
```

!!!

!!!

That was even faster!

There is a lot more that can go into semantic search. Stay tuned for a follow up post on hybrid search and re-ranking.

If you have any questions, or just have an idea on how to make PostgresML better, we'd love to hear from you in our [Discord](https://discord.com/invite/DmyJP3qJ7U). We’re open source, and welcome contributions from the community, especially when it comes to the rapidly evolving ML/AI landscape.

> ## Closing thoughts / why PostgresQL?

Reviewer 2: ~PostgresQL~ PostgreSQL
